File: src/main/java/model/pieces/Rook.java
package model.pieces;

import model.moves.Move;
import model.util.Colour;
import model.util.Offset;

public class Rook extends DirectionPiece {
    private static final Offset[] OFFSETS = {
            Offset.UP,
            Offset.LEFT,
            Offset.RIGHT,
            Offset.DOWN
    };

    /**
     * The number of moves that have been applied to this piece. Used to know if the piece has moved
     */
    private int numberOfAppliedMoves = 0;

    public Rook(Colour colour) {
        super(colour);
    }

    @Override
    int getUnicodeWhite() {
        return 9814;
    }

    @Override
    int getUnicodeBlack() {
        return 9820;
    }

    @Override
    Offset[] getDirections() {
        return OFFSETS;
    }

    @Override
    public int getUnsignedValue() {
        return 5;
    }

    @Override
    public void notifyMoveComplete(Move move) {
        numberOfAppliedMoves += 1;
    }

    @Override
    public void notifyMoveUndo(Move move) {
        numberOfAppliedMoves -= 1;
    }

    boolean hasMoved() {
        return numberOfAppliedMoves != 0;
    }

    @Override
    String getName() {
        return "Rook";
    }
}

File: src/main/java/model/pieces/Rook.kt
package model.pieces

import model.moves.Move
import model.util.Colour
import model.util.Offset

internal class Rook(colour: Colour) : DirectionPiece(colour) {

    /**
     * The number of moves that have been applied to this piece. Used to know if the piece has moved
     */
    private var numberOfAppliedMoves = 0

    override val unicodeWhite: Int
        get() = 9814

    override val unicodeBlack: Int
        get() = 9820

    override val directions: Array<Offset>
        get() = OFFSETS

    override val unsignedValue: Int
        get() = 5

    override val name: String
        get() = "Rook"

    override fun notifyMoveComplete(move: Move) {
        numberOfAppliedMoves += 1
    }

    override fun notifyMoveUndo(move: Move) {
        numberOfAppliedMoves -= 1
    }

    internal fun hasMoved(): Boolean {
        return numberOfAppliedMoves != 0
    }

    companion object {
        private val OFFSETS = arrayOf(Offset.UP, Offset.LEFT, Offset.RIGHT, Offset.DOWN)
    }
}
